### DISCUSSION TITLE
[small feature] Add \e string escape sequence

### DISCUSSION THREAD

**Discussion by Unknown6656** [dotnet/csharplang#7394] [Language Ideas]
## Background

We currently have the following escape sequences in C# strings:

- `\'` - single quote, needed for character literals
- `\"` - double quote, needed for string literals
- `\\` - backslash
- `\0` - Unicode character 0
- `\a` - Alert (character 7)
- `\b` - Backspace (character 8)
- `\f` - Form feed (character 12)
- `\n` - New line (character 10)
- `\r` - Carriage return (character 13)
- `\t` - Horizontal tab (character 9)
- `\v` - Vertical tab (character 11)
- `\uxxxx` - Unicode escape sequence for character with hex value xxxx
- `\xn[n][n][n]` - Unicode escape sequence for character with hex value nnnn (variable length version of \uxxxx)
- `\Uxxxxxxxx` - Unicode escape sequence for character with hex value xxxxxxxx (for generating surrogates)

However, as I am working very often with [VT100/ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code), I do miss a short escape sequence for the character `0x1b`, which is needed all the time in ANSI escape sequences.
Currently, I often write myself writing stuff like:

```csharp
Console.Write("\u001b[4m\u001b[6m\u001b[38;2;205;20;60mThis is a bold, blinking, underlined, red text.\u001b[0mThis is a regular text.");
```
![image](https://github.com/dotnet/csharplang/assets/8807985/0ee72913-a977-4f40-9d7f-e991650be098)<br/>
![image](https://github.com/dotnet/csharplang/assets/8807985/02c6d733-d0bb-4171-817c-8fd7c9075f48)

## Actual Feature Request

I know that it only saves a couple of characters, but I'd really love to see `\e` as a shorthand for `\u001b` as a new escape sequence for C# strings, in order to enhance readability in the context of VT100/ANSI escape codes:

```csharp
Console.Write("\e[4m\e[6m\e[38;2;205;20;60mThis is a bold, blinking, underlined, red text.\e[0mThis is a regular text.");
```

## Implementation Cost

Hopefully, it should be low, as the internal translation would be identical to how e.g. `\n` gets translated to `\u000a`.

## Alternatives

Alternatively, one can always use the following:
```csharp
const char ESC = (char)0x1b;
Console.Write($"{ESC}[4m{ESC}[6m{ESC}[38;2;205;20;60mThis is a bold, blinking, underlined, red text.{ESC}[0mThis is a regular text.");
```
or:
```csharp
static string escape(string code) => $"\u001b[{code}m";
Console.Write($"{escape("4")}{escape("6")}{escape("38;2;205;20;60")}This is a bold, blinking, underlined, red text.{escape("0")}This is a regular text.");
```

**Comment by CyrusNajmabadi** [comment]
I'd champion a proposal here.  Having done the raw-strings-work, i know this would be a trivial thing to implement on compiler and IDE side.

**Reply by CyrusNajmabadi** [reply]
Before going into a spec on this though.  I'd like to know from you @Unknown6656  (and @jnm2 ) if there are other common cases that would warrant something.  I'd want to roll everything together if possible.

**Reply by jnm2** [reply]
Unless someone else is looking for a chance to get involved, I'll write up a proposal.

**Reply by CyrusNajmabadi** [reply]
@Unknown6656 We'd be happy to let you write up the spec proposal here.  If not, we're happy to take it on for you.

That said, i want to be clear about what happens next.  First, the team is entirely slammed on C#12 work and has a huge workload ahead of it for C#13.  When we open up the working group for new proposals, I would bring this along (as well as possibly any others that might go tightly with this) to get a team decision on if we would actually take it or not.  If so, then we'd work out the schedule/timeline to implement (though given how trivial this is, that would hopefully nto be an issue).

So, in other words, while this is me (a team member) expressing that I will support this going forward, it is not a guarantee that it will actually be approved and incorporated.  I just want to set expectations there so there is understanding if this doesn't happen. 

Thanks!

**Reply by Unknown6656** [reply]
@jnm2 I'd love to help you on that. However, I've never written a correct/formal language feature proposal here. I'll give it a try and you can support me if you want to.

@CyrusNajmabadi regarding the first comment: unfortunately, I do not know any real other use cases besides VT100/ANSI escape sequences.

regarding the second comment: I'll happily write a proposal, however, as mentioned, I've never really done that for this repository. So you'll most definitely have to review it to a certain degree. I'm thankful that you did some expectation management there, and I will definitely keep that in mind.
_(However, seeing as I've been waiting for discriminated unions ever since C#6, I will have no hard feelings if the implementation of `\e` gets delayed for a couple of language iterations 😜)_

**Reply by jnm2** [reply]
Another use case I've been involved with is low-level device communication, such as printer ESC/POS commands.

**Reply by Unknown6656** [reply]
@jnm2 good point. I'm currently writing the porposal, so I'll include that. feel free to drop some references here and I'll include them :)

**Reply by CyrusNajmabadi** [reply]
> However, I've never written a correct/formal language feature proposal here. I'll give it a try and you can support me if you want to.

First, a reasonable way to do this (For a small tweak feature like this) would be to use this link:

![image](https://github.com/dotnet/csharplang/assets/4564579/fe38a758-9bbe-4169-9cd9-19a6f2ce0d46)

Which will give you a template (also make a note saying Cyrus approved you opening this issue).

Second, try to make the spec include a 'diff' of the rules of the current language.  For that, it would be basically specifying the changes you woudl make to: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/lexical-structure.md#6455-character-literals

you can use the ```` ```diff ```` tag for this purpose in the spec.  

--

I would also nto go overboard at first.  Just write up a simple first sketch.  We can then help you revise as necessary.  This will help avoid a lot of churn, and will hopefully save time for you :)

**Reply by jnm2** [reply]
The #csharp-lang-design channel at https://aka.ms/dotnet-discord is a great place for asking questions, too.

**Reply by jnm2** [reply]
@Unknown6656 Also see https://learn.microsoft.com/en-us/windows/uwp/devices-sensors/epson-esc-pos-with-formatting or https://reference.epson-biz.com/modules/ref_escpos/

**Comment by Unknown6656** [comment]
Closing in favor of #7400.
THanks @CyrusNajmabadi  and @jnm2  for your input!
